src/DeterministicNoiseSimulator.cpp

```diff
@@ -106,169 +122,214 @@ std::map<std::string, double> DeterministicNoiseSimulator::DeterministicSimulate
-                    if (classic_values[i] != (expValue % 2)) {
-                        conditionTrue = false;
-                        break;
-                    } else {
-                        expValue = expValue >> 1u;
-                    }
-                }
-                dd_op   = dd::getDD(classic_op->getOperation(), dd);
-                targets = classic_op->getOperation()->getTargets();
-                assert(targets.size() == 1);
-                controls = classic_op->getOperation()->getControls();
-                if (!conditionTrue) {
-                    continue;
-                }
+                throw std::runtime_error("Classical controlled operations are not supported.");
+                //                // Check if the operation is controlled by a classical register
+                //                auto* classic_op    = dynamic_cast<qc::ClassicControlledOperation*>(op.get());
+                //                bool  conditionTrue = true;
+                //                auto  expValue      = classic_op->getExpectedValue();
+                //
+                //                for (unsigned int i = classic_op->getControlRegister().first;
+                //                     i < classic_op->getControlRegister().second; i++) {
+                //                    if (classic_values[i] != (expValue % 2)) {
+                //                        conditionTrue = false;
+                //                        break;
+                //                    } else {
+                //                        expValue = expValue >> 1u;
+                //                    }
+                //                }
+                //                if (!conditionTrue) {
+                //                    continue;
+                //                }
+                //                dd_op    = dd::getDD(classic_op->getOperation(), dd);
+                //                targets  = classic_op->getOperation()->getTargets();
+                //                controls = classic_op->getOperation()->getControls();
             } else {
-                dd_op   = dd::getDD(op.get(), dd);
-                targets = op->getTargets();
-                assert(targets.size() == 1);
+                dd_op    = dd::getDD(op.get(), dd);
+                targets  = op->getTargets();
                 controls = op->getControls();
             }
+
             // Applying the operation to the density matrix
-            qc::MatrixDD tmp0 = dd->multiply(dd->multiply(dd_op, density_root_edge), dd->conjugateTranspose(dd_op));
-            dd->incRef(tmp0);
+            auto tmp0 = dd->conjugateTranspose(dd_op);
+            auto tmp1 = dd->multiply(reinterpret_cast<dEdge&>(density_root_edge), reinterpret_cast<dEdge&>(tmp0), 0, false);
+            auto tmp2 = dd->multiply(reinterpret_cast<dEdge&>(dd_op), reinterpret_cast<dEdge&>(tmp1), 0, use_density_matrix_type);
```

Comment:
Application of an operation to a density matrix needs to be encapsulated as a function in the DD package just as matrix-vector and matrix-matrix multiplication are.

Reply:
done. Added 
> dEdge applyOperationToDensity(dEdge& e, mEdge& operation, [[maybe_unused]] bool generateDensityMatrix = false)
in the ddPackage